### DISCUSSION TITLE
SQLAlchemy core - stable production setup

### DISCUSSION THREAD

**Discussion by mdczaplicki** [sqlalchemy/sqlalchemy#10146] [Usage Questions]
I'd like to ask you how do you connect to the database using SQLAlchemy core in production environments? It's a general question about your setups.
In our company - we use SA core for few years already. First we connected to PostgreSQL hosted on DigitalOcean - it had many flaws, it wasn't as stable as we wanted for it to be, and we didn't have enough control over the database's resources.
We have migrated to GCP. Again PostgreSQL. PSQL 14.7 as of now. The database seems to be doing fine in terms of resources, we don't overload it, yet we encounter a lot of:
`ConnectionResetError (asyncpg/connect_utils.py in _create_ssl_connection at line 692)`
or
`ConnectionError (asyncpg/connect_utils.py in _create_ssl_connection at line 692)`
We have high availability present for our PostgreSQL instance.

This is how we create an engine in our code:
```
    engine = create_async_engine(
        connection_string,
        connect_args={"server_settings": {"application_name": app_name}},
        pool_size=abs(math.ceil(worker_count / 10)) or 1,
        max_overflow=abs(worker_count) * 2,
        pool_pre_ping=True,
        pool_recycle=3600,  # after 1h a connection will be discarded & recreated
        **kwargs,
    )
```
`pool_recycle` is a recent idea, but didn't work as expected. We rely on connection pool management that comes from SA.

We tend to open connections to the database when they're needed e.g., before getting/saving some data from the database. Sometimes, when we want multiple queries to be executed in on transaction - we open one connection on higher abstraction level and downstream it to query-functions. Right after we don't need a connection - we close it.

What are we missing in our setup to make it more stable? We're running it for few years already, it's quite stable, but I feel like we are missing something to make it more stable. Any thoughts?

**Comment by zzzeek** [comment]
When did these connection reset errors appear?   were they related to:  1. the driver 2. the use of SSL 3. a specific DB platform ?   What's the complete stack trace that illustrates the complete operation in progress?

What is not working with pool_recycle?  the error you show, if the asyncpg function is aptly named, seems to be occurring when a connection first occurs.         

have you considered using a more industrial strength load balancer like PGBouncer ?

**Comment by CaselIT** [comment]
Hi,

There aren't much details provided to suggest what may be going on. Even for the mentioned errors there is no message provided.

Just a note that there isn't much use to set pool recycle if you also set pool pre ping. Should not cause the seen issue though

**Comment by mdczaplicki** [comment]
This is a part of the stack that I can share with you:
![stack](https://github.com/sqlalchemy/sqlalchemy/assets/9108586/6fe18adf-cb96-4629-ad41-055c070ab082)

@zzzeek, I'm not saying that `pooly_recycle` doesn't work, or anything (I'm not the one that introduced it). I'm just saying that it didn't mitigate that problem. It could have mitigated a different problem though, not sure. Doubt that it's relevant in here. Especially that @CaselIT, already told that `pool_recycle` doesn't go along with `pool_pre_ping`.

By the amount of the errors - I doubt that it happens when a connection first occurs.
<img width="325" alt="image" src="https://github.com/sqlalchemy/sqlalchemy/assets/9108586/9702cc49-f7b4-4a39-a7a2-9b5df3e08b11">

PGBouncer is something that we do not have, but is there anything else you can point me to?

**Reply by zzzeek** [reply]
I'm going to make a guess that your issue might be that your program is overly CPU bound, and your asyncpg wait_for is timing out because the program is too busy with other CPU-bound tasks in order to attend to PostgreSQL's SSL challenge in time.    unfortunately I'm not able to find background information on what this timeout might default towards or how it may be affected.

**Reply by zzzeek** [reply]
note that "socket message not responded towards in a timely manner due to CPU bound issues " is a problem that is specific to cooperative multitasking approaches to non-blocking IO, namely asyncio as well as greenlet/eventlet.     PostgreSQL makes the SSL challenge available and it is certainly the kind of thing it's going to cut off after some short delay - under asyncio, the part of the program that would respond to this challenge will not be invoked until whatever else the program is doing is awaited.

**Reply by zzzeek** [reply]
anyway....PGBouncer could potentially resolve this since it would be handling making these SSL connections.  however, your client would be making plaintext connections to PGBouncer.  if there were some SSL client config on PGBouncer, you'd have the same problem unless you could make for a longer timeout at that level.

**Reply by zzzeek** [reply]
you might want to try authentication_timeout at https://www.postgresql.org/docs/current/runtime-config-connection.html but that's set to one minute....I suspect this might be more of an openssl thing at a deeper level

**Reply by mdczaplicki** [reply]
I've just checked and we are not CPU-bound. It happens in many of our services. Most of them run at around 20-30% of CPU utilization.
Is the SSL challenge required? We do not use SSL, this is internal communication without encryption.

I guess we'll try out PGBouncer.

**Reply by CaselIT** [reply]
Have you tried using another dbapi? 
Also since all the issue are from the dbapi it may make sense asking for guidance also from them

**Reply by mdczaplicki** [reply]
No, we haven't. I doubt that we will though.

That's a good idea. We are onto something, so I'll wait a little bit.

**Comment by mdczaplicki** [comment]
Turns out that the root cause of those errors was an SQL-proxy that we have when we connect to the database server. That proxy was under-resources (CPU-wise), so @zzzeek, you were right. I was looking into wrong place.

Thank you guys!
